### PR TITLE
Don't encode/decode image when app will terminated

### PR DESCRIPTION
**Pull request by kinarobin**
### New Pull Request Checklist

* [x] I have read and understood the [CONTRIBUTING guide](https://github.com/rs/SDWebImage/blob/master/.github/CONTRIBUTING.md)
* [x] I have read the [Documentation](http://cocoadocs.org/docsets/SDWebImage/)
* [x] I have searched for a similar pull request in the [project](https://github.com/rs/SDWebImage/pulls) and found none

* [x] I have updated this branch with the latest master to avoid conflicts (via merge from master or rebase)
* [x] I have added the required tests to prove the fix/feature I am adding
* [x] I have updated the documentation (if necessary)
* [x] I have run the tests and they pass
* [x] I have run the lint and it passes (`pod lib lint`)

This merge request fixes / refers to the following issues: ...

Maybe relate #3049 #2229

### CODE REVIEW

File: SDWebImage/Core/SDImageIOAnimatedCoder.m [hunk 1]
```diff
@@ -152,11 +186,17 @@ + (NSUInteger)imageLoopCountWithSource:(CGImageSourceRef)source {
 }
 
 + (NSTimeInterval)frameDurationAtIndex:(NSUInteger)index source:(CGImageSourceRef)source {
+    NSTimeInterval frameDuration = 0.1;
+    
+    // Earily return when application will be terminated.
+    if (self.willTerminated) {
```
Comment (dreampiggy) [inline SDWebImage/Core/SDImageIOAnimatedCoder.m:192]: Don't add guard on this method. Wrong design.

This is a helper method, add to the caller.

File: SDWebImage/Core/SDImageIOAnimatedCoder.m [hunk 2]
```diff
@@ -29,6 +29,9 @@ @interface SDImageIOCoderFrame : NSObject
 @implementation SDImageIOCoderFrame
 @end
 
+BOOL willTerminated_;
+NSLock *terminatedLock_;
```
Comment (dreampiggy) [inline SDWebImage/Core/SDImageIOAnimatedCoder.m:33]: Don't use global variable. Use instance property instead.

**Review by dreampiggy — CHANGES_REQUESTED** [review]

File: SDWebImage/Core/SDImageIOAnimatedCoder.m (hunk 1 above)
Comment (kinarobin) [reply SDWebImage/Core/SDImageIOAnimatedCoder.m]: Done.

File: SDWebImage/Core/SDImageIOAnimatedCoder.m (hunk 2 above)
Comment (kinarobin) [reply SDWebImage/Core/SDImageIOAnimatedCoder.m]: In most cases, instances of  'SDImageIOAnimatedCoder'  are not created,  so there seems to be no way to use instance property variables in this place

Comment (dreampiggy) [reply SDWebImage/Core/SDImageIOAnimatedCoder.m]: OK

**Review by dreampiggy — APPROVED** [review]

**Pull request merged**